core/commands/dag/dag.go

```diff
@@ -241,3 +278,421 @@ var DagResolveCmd = &cmds.Command{
+	err   error
+}
+
+var DagImportCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Import the contents of .car files",
+		ShortDescription: `
+'ipfs dag import' parses .car files and adds all objects to the blockstore.
+By default, after all car files have been processed, an attempt is made to
+pin each individual root specified in the car headers, before GC runs again.
+`,
+	},
+	Arguments: []cmds.Argument{
+		cmds.FileArg("path", true, true, "The path of a .car file.").EnableStdin(),
+	},
+	Options: []cmds.Option{
+		cmds.BoolOption(progressOptionName, "p", "Display progress on CLI.").WithDefault(true),
+		cmds.BoolOption(silentOptionName, "No output."),
+		cmds.BoolOption(pinRootsOptionName, "Pin optional roots listed in the .car headers after importing.").WithDefault(true),
+	},
+	PreRun: func(req *cmds.Request, env cmds.Environment) error {
+		if silent, _ := req.Options[silentOptionName].(bool); silent {
+			req.Options[progressOptionName] = false
+		}
+
+		return nil
+	},
+	Type: CarImportOutput{},
+	Run: func(req *cmds.Request, res cmds.ResponseEmitter, env cmds.Environment) error {
+
+		node, err := cmdenv.GetNode(env)
+		if err != nil {
+			return err
+		}
+
+		api, err := cmdenv.GetApi(env, req)
+		if err != nil {
+			return err
+		}
+
+		// on import ensure we do not reach out to the network for any reason
+		// if a pin based on what is imported + what is in the lockstore
+		// isn't possible: tough luck
+		api, _ = api.WithOptions(options.Api.Offline(true))
+
+		// grab a gc lock so that regardless of the size of the streamed-in cars
+		// nothing will disappear on us before we had a chance to pin the roots
+		// at the very end
+		// This is especially important for use cases like dagger:
+		//    ipfs dag import $( ... | ipfs-dagger --stdout=carfifos )
+		//
+		// !!!NOTE!!! - We take .GCLock, *not* .PinLock. The way the current stack is
+		// built makes it impossible to take both a GC and a Pinning lock, regardless
+		// of orer. Doing so can potentially deadlock if another thread starts GC
+		// ( this is not theoretical - it's been reliably simulated with the insertion
+		// of strategic sleeps()s and hand-triggering GC )
+		//
+		// See api.Pin().Add() vs node.Pinning.Pin() comment further down
+		unlocker := node.Blockstore.GCLock()
+		defer func() { unlocker.Unlock() }()
+
+		doPinRoots, _ := req.Options[pinRootsOptionName].(bool)
+		showProgress, _ := req.Options[progressOptionName].(bool)
+
+		retCh := make(chan importResult, 1)
+		var counts ObjectCounts
+		go importWorker(req, &res, &api, &counts, retCh)
+
+		var progressTicker *time.Ticker
+		if showProgress {
+			progressTicker = time.NewTicker(500 * time.Millisecond)
+			defer progressTicker.Stop()
+		}
+
+		var roots *expectedRootsSeen
+
+	ImportLoop:
+		for {
+			select {
+			case ret := <-retCh:
+				if ret.err != nil {
+					return ret.err
+				}
+				roots = ret.roots
+				if progressTicker != nil {
+					progressTicker.Stop()
```

Comment:
This is where I emit ( coupe lines down ) a trailing "\n". The thinking was to avoid a garbled CLI display by:
- making sure no further updates take place
- do one last udate
- newline so the "\r" is neutralised

Am I missing something?

Reply:
I was missing... a lot of things, see https://github.com/ipfs/go-ipfs/pull/7011#discussion_r396024829 and https://github.com/ipfs/go-ipfs/pull/7011/commits/67642d8d5e7a5b56bd8e1b5f3c3e4f64f930aac9